rfc/terraform-provider.md

```diff
@@ -0,0 +1,92 @@
+- Infrastructure provisioning should explicitly **not** be performed by the Constellation Terraform provider.
+- The User needs to receive some sort of Access Token to the Cluster (e.g. a Kubeconfig, whether long- or short-lived)
+when applying a configuration containing the provider and the resources listed below.
+
+## Terraform Configuration Layout for the Constellation Resource
+
+This resembles an examplary configuration of a Constellation cluster through Terraform. While the exact naming and struct layout
+of the individual components may change until the implementation, the structure of the overall configuration looks like this:
+
+```hcl
+terraform {
+  required_providers {
+    constellation = {
+      source  = "tbd/constellation"
+      version = "2.13.0"
+    }
+  }
+}
+
+provider "constellation" { }
+
+resource "constellation_cluster" "foo" {
+    uid = "bar"
+    name = "baz"
+    image = data.constellation_image.ref # or provide manually crafted values
+    kubernetes_version = "v1.27.6"
+    constellation_microservice_version = "lockstep" # or pinned to a specific version
+    debug = false
+    init_endpoint = "10.10.10.10" # should use public ip of LB resource, ideally also provisioned through TF
+    kubernetes_api_endpoint = "10.10.10.10" # should use public ip of LB resource, ideally also provisioned through TF
+    extra_microservices = {
+        csi_driver = true
+        # + more
+        # possiblly also constellation microservices with version and maybe service options,
+        # which would make constellation_microservice_version obsolete.
+        # exact API TBD
+    }
+    master_secret = "foo"
+    init_secret = "bar" # maybe derive from master_secret
+    network_config = {
+        # TBD
+        # should contain CIDRs for pod network, service cidr, node network... for Cilium
+        # and in-cluster Kubernetes API endpoint, e.g. for Kubelets
+    }
+    attestation = data.constellation_attestation.attestation # or provide manually crafted values
+}
+
+# constellation_cluster provides:
+# constellation_cluster.foo.kubeconfig
+# constellation_cluster.foo.owner_id
+# constellation_cluster.foo.cluster_id
+
+
+data "constellation_attestation" "foo" {
+    attestation_variant = "GCP-SEV-ES"
+    image_version = "v2.13.0" # or "lockstep"
+    maa_url = "https://www.example.com" # optional, for Azure only
+}
+
+# constellation_attestation provides:
+# data.constellation_attestation.foo.measurements
+# data.constellation_attestation.foo.attestation
+
+
+data "constellation_image" "foo" {
+    image_version = "v2.21.0" # or "lockstep"
+    attestation_variant = "GCP-SEV-ES"
+    csp = "GCP"
+    region = "us-central1" # optional, for AWS only
+}
+
+# constellation_image provides:
+# constellation_image.foo.reference (CSP-specific image reference)
+```
+
+The `constellation_cluster` resource is the main resource implemented by the provider.
+It resembles a Constellation cluster with a specific configuration.
```

Comment:
>  What does resemble mean? Is it not the same as a CLI created cluster?

It is, but I intended to say that we have the Terraform resource declaration on one side, and the actual cloud resources & data on the other side.

> What is the behavior of recreate?

Recreate means destroy and create, performed in that order. (i.e. you get a new cluster)

> Is it possible to perform a new init on an initialized cluster?

No

Reply:
How would we enforce that the infra is recreated when the master secret changes? Atm, the infra is only related to the secret through a tag (only checked AWS). I don't think this would force a recreate. So I understand that we would design the infra module to force a recreate when it changes?

Reply:
> It is, but I intended to say that we have the Terraform resource declaration on one side, and the actual cloud resources & data on the other side.

I would reword the sentence then. "The resource declares the provisioning / management of a Constellation cluster on the provided infrastructure."

Reply:
> How would we enforce that the infra is recreated when the master secret changes?

I think we should throw an error [like so](https://developer.hashicorp.com/terraform/plugin/framework/migrating/attributes-blocks/force-new#framework). Changing the underlying infrastructure is not necessary, the disks of the machines can just be overwritten with new images, leaving them in a "fresh" state. The only thing we need to warn the user about are previously created persistent volumes, which are encrypted with a secret derived from the old mastersecret and thus cannot be accessed with the new mastersecret.

Reply:
Pls add this to the RFC.